hr_contract_reference/models/hr_contract.py

```diff
@@ -17,30 +18,21 @@
 #    You should have received a copy of the GNU Affero General Public License
 #    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #
-#
+##############################################################################
 
-from openerp.osv import fields, orm
+from openerp import models, fields, api
 
 
-class hr_contract(orm.Model):
-
+class hr_contract(models.Model):
     _inherit = 'hr.contract'
 
-    _columns = {
-        'name': fields.char(
-            'Contract Reference',
-            size=32,
-            required=False,
-            readonly=True,
-        ),
-    }
-
-    def create(self, cr, uid, vals, context=None):
+    name = fields.Char('Contract Reference', required=False,
+            readonly=True, copy=False)
 
-        cid = super(hr_contract, self).create(cr, uid, vals, context)
-        if cid:
-            ref = self.pool.get('ir.sequence').next_by_code(
-                cr, uid, 'contract.ref', context=context)
-            self.pool.get('hr.contract').write(
-                cr, uid, cid, {'name': ref}, context=context)
-        return cid
+    @api.model
+    def create(self, vals):
+        contract = super(hr_contract, self).create(vals)
```

Comment:
Put before and inject in vals the sequence to avoid 2 operations. Check also if the value is /, and then get the sequence. Any other value, leave it as is. This is for make it easy to work with methods that writes directly a contract number.